includes/core-changes/corefx/5.0/lastindexof-improved-handing-of-empty-values.md

```diff
@@ -0,0 +1,59 @@
+### LastIndexOf has improved handling of empty search strings
+
+<xref:System.String.LastIndexOf%2A?displayProperty=nameWithType> and related APIs now return correct values when searching for a zero-length (or zero-length equivalent) substring within a larger string.
+
+#### Change description
+
+In .NET Framework and .NET Core 1.0 - 3.1, <xref:System.String.LastIndexOf%2A?displayProperty=nameWithType> and related APIs might return an incorrect value when the caller searches for a zero-length substring.
+
+```csharp
+Console.WriteLine("Hello".LastIndexOf("")); // prints '4' (incorrect)
+
+ReadOnlySpan<char> span = "Hello";
+Console.WriteLine(span.LastIndexOf("")); // prints '0' (incorrect)
+```
+
+Starting with .NET 5.0, these APIs return the correct value for `LastIndexOf`.
+
+```csharp
+Console.WriteLine("Hello".LastIndexOf("")); // prints '5' (correct)
+
+ReadOnlySpan<char> span = "Hello";
+Console.WriteLine(span.LastIndexOf("")); // prints '5' (correct)
+```
+
+In these examples, `5` is the correct answer because `"Hello".Substring(5)` and `"Hello".AsSpan().Slice(5)` both produce an empty string, which is trivially equal to the empty substring that is sought.
+
+#### Reason for change
+
+This change was part of an overall bug fixing effort around string handling for .NET 5. It also helps unify our behavior between Windows and non-Windows platforms. For more information, see [dotnet/runtime#13383](https://github.com/dotnet/runtime/issues/13383) and [dotnet/runtime##13382](https://github.com/dotnet/runtime/issues/13382).
+
+#### Version introduced
+
+5.0
+
+#### Recommended action
+
+You don't need to take any action. The .NET 5.0 runtime provides the new behaviors automatically.
```

Comment:
I don't think this is correct.

---

```csharp
private static void M(string s, string s2)
{
    var index = s.LastIndexOf(s2);
    if (index > -1) Console.WriteLine(s[index]);
}
```

The above code was valid before this breaking change.

But with this breaking change, the developer should check the against s.Length or check that s is not empty before trying to access the index.

See #44376

Reply:
This code was never valid. See my response at https://github.com/dotnet/runtime/issues/44376.

Reply:
@GrabYourPitchforks I still think this may need developer action in some cases, see a possible example from dotnet/aspnetcore in https://github.com/dotnet/runtime/issues/44376#issuecomment-723549583.